R/mock-session.R

```diff
@@ -0,0 +1,283 @@
+      out <- .createOutputWriter(self)
+      class(out) <- "shinyoutput"
+      self$output <- out
+
+      # Create a read-only copy of the inputs reactive.
+      self$input <- .createReactiveValues(private$.input, readonly = TRUE)
+    },
+    onFlush = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushCBs$register(fun))
+      } else {
+        dereg <- private$flushCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onFlushed = function(fun, once) {
+      if (!isTRUE(once)) {
+        return(private$flushedCBs$register(fun))
+      } else {
+        dereg <- private$flushedCBs$register(function() {
+          dereg()
+          fun()
+        })
+        return(dereg)
+      }
+    },
+    onEnded = function(sessionEndedCallback) {
+      private$endedCBs$register(sessionEndedCallback)
+    },
+
+    isEnded = function(){ private$closed },
+    isClosed = function(){ private$closed },
+    close = function(){ private$closed <- TRUE },
+
+    #FIXME: this is wrong. Will need to be more complex.
+    cycleStartAction = function(callback){ callback() },
+
+    # Needed for image rendering. Base64-encode the given file.
+    fileUrl = function(name, file, contentType='application/octet-stream') {
+      bytes <- file.info(file)$size
+      if (is.na(bytes))
+        return(NULL)
+
+      fileData <- readBin(file, 'raw', n=bytes)
+      b64 <- rawToBase64(fileData)
+      return(paste('data:', contentType, ';base64,', b64, sep=''))
+    },
+
+    setInputs = function(...) {
+      vals <- list(...)
+      # TODO: is there really not a way to access `names` from inside an lapply?
+      lapply(names(vals), function(k){
+        v <- vals[[k]]
+        private$.input$set(k, v)
+      })
+
+      private$flush()
+    },
+
+    .scheduleTask = function(millis, callback) {
+      id <- private$timer$schedule(millis, callback)
+
+      # Return a deregistration callback
+      function() {
+        invisible(private$timer$unschedule(id))
+      }
+    },
+    elapse = function(millis) {
+      msLeft <- millis
+
+      while (msLeft > 0){
+        t <- private$timer$timeToNextEvent()
+
+        if (is.infinite(t) || t <= 0 || msLeft < t){
+          # Either there's no good upcoming event or we can't make it to it in the allotted time.
+          break
+        }
+        msLeft <- msLeft - t
+        private$timer$elapse(t)
+
+        # timerCallbacks must run before flushReact.
+        private$timer$executeElapsed()
+        private$flush()
+      }
+
+      private$timer$elapse(msLeft)
+
+      # TODO: needed? We're guaranteed to not have anything to run given the above loop, right?
+      private$timer$executeElapsed()
+      private$flush()
+    },
+
+    .now = function() {
+      # Contract is to return Sys.time, which is seconds, not millis.
+      private$timer$getElapsed()/1000
+    },
+
+    defineOutput = function(name, value, label) {
```

Comment:
There are a bunch of tricky things that the real `defineOutput` does that I think the the mock version will also need to do:

* Reactive domain: There needs to be a reactive domain in order for the observer to be automatically cleaned up. (I'm not sure if that's already happening or not.)
* Overwriting an output should destroy the previous one.
* I think that `value` should be renamed to `func`, for consistency with the real `defineOutput`.
* Might not be worth addressing right now (maybe in a separate PR), but we'll probably want to wrap the call to `func()` in ` private$withCurrentOutput()`. This is so the `getCurrentOutputInfo()` function will work. That function allows code in the output, or in any reactives that it calls, to introspect and find out information about the output that's being executed. The information is generated by https://github.com/rstudio/shiny/blob/005295f/R/shiny.R#L1298-L1337

Reply:
Spoke live. Points 2, 3 are handled in recent commits. Point 1 will fall to the caller (in a subsequent PR). Point 4 will be a follow-on PR to come later. Will prioritize as necessary based on feedback of people who hit that.